File: framec_tests/golang/state_params/state_params.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files
package state_params

import "golang/framelang"


func NewStateParams() StateParams {
    m := &stateParamsStruct{}
    
    // Validate interfaces
    var _ StateParams = m
    var _ StateParams_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewStateParamsCompartment(StateParamsState_Init)
    
    // Override domain variables.
    m.param_log = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type StateParamsState uint

const (
    StateParamsState_Init StateParamsState = iota
    StateParamsState_Split
    StateParamsState_Merge
)

type StateParams interface {
    Next() 
    Prev() 
    Log() 
}

type StateParams_actions interface {
    got_param(name string,val int) 
}


type stateParamsStruct struct {
    _compartment_ *StateParamsCompartment
    _nextCompartment_ *StateParamsCompartment
    param_log []string
}

//===================== Interface Block ===================//

func (m *stateParamsStruct) Next()  {
    e := framelang.FrameEvent{Msg:"Next"}
    m._mux_(&e)
}

func (m *stateParamsStruct) Prev()  {
    e := framelang.FrameEvent{Msg:"Prev"}
    m._mux_(&e)
}

func (m *stateParamsStruct) Log()  {
    e := framelang.FrameEvent{Msg:"Log"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *stateParamsStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case StateParamsState_Init:
        m._StateParamsState_Init_(e)
    case StateParamsState_Split:
        m._StateParamsState_Split_(e)
    case StateParamsState_Merge:
        m._StateParamsState_Merge_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *stateParamsStruct) _StateParamsState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Next":
        compartment := NewStateParamsCompartment(StateParamsState_Split)
        compartment.StateArgs["val"] = 1
        
        m._transition_(compartment)
        
        return
    }
}

func (m *stateParamsStruct) _StateParamsState_Split_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Next":
        compartment := NewStateParamsCompartment(StateParamsState_Merge)
        compartment.StateArgs["left"] = m._compartment_.StateArgs["val"].(int)
        
        compartment.StateArgs["right"] = m._compartment_.StateArgs["val"].(int) + 1
        
        m._transition_(compartment)
        
        return
    case "Prev":
        compartment := NewStateParamsCompartment(StateParamsState_Merge)
        compartment.StateArgs["left"] = m._compartment_.StateArgs["val"].(int) + 1
        
        compartment.StateArgs["right"] = m._compartment_.StateArgs["val"].(int)
        
        m._transition_(compartment)
        
        return
    case "Log":
        m.got_param("val",(m._compartment_.StateArgs["val"].(int)))
        
        return
    }
}

func (m *stateParamsStruct) _StateParamsState_Merge_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Next":
        compartment := NewStateParamsCompartment(StateParamsState_Split)
        compartment.StateArgs["val"] = m._compartment_.StateArgs["left"].(int) + m._compartment_.StateArgs["right"].(int)
        
        m._transition_(compartment)
        
        return
    case "Prev":
        compartment := NewStateParamsCompartment(StateParamsState_Split)
        compartment.StateArgs["val"] = m._compartment_.StateArgs["left"].(int) * m._compartment_.StateArgs["right"].(int)
        
        m._transition_(compartment)
        
        return
    case "Log":
        m.got_param("left",(m._compartment_.StateArgs["left"].(int)))
        m.got_param("right",(m._compartment_.StateArgs["right"].(int)))
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *stateParamsStruct) _transition_(compartment *StateParamsCompartment) {
    m._nextCompartment_ = compartment
}

func (m *stateParamsStruct) _do_transition_(nextCompartment *StateParamsCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *stateParamsStruct) got_param(name string,val int)  {}

********************************************************/

//=============== Compartment ==============//

type StateParamsCompartment struct {
    State StateParamsState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewStateParamsCompartment(state StateParamsState) *StateParamsCompartment {
    c := &StateParamsCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/state_params/state_params.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class StateParams:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = StateParamsCompartment('__stateparams_state_Init', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = [None]
        
        # Initialize domain
        
        self.param_log  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def Next(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Next",None)
        self.__kernel(__e)
        return self.return_stack.pop(-1)
    
    def Prev(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Prev",None)
        self.__kernel(__e)
        return self.return_stack.pop(-1)
    
    def Log(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Log",None)
        self.__kernel(__e)
        return self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $Init
    
    def __stateparams_state_Init(self, __e, compartment):
        if __e._message == "Next":
            next_compartment = None
            next_compartment = StateParamsCompartment('__stateparams_state_Split', next_compartment)
            next_compartment.state_args["val"] = 1
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $Split
    
    def __stateparams_state_Split(self, __e, compartment):
        if __e._message == "Next":
            next_compartment = None
            next_compartment = StateParamsCompartment('__stateparams_state_Merge', next_compartment)
            next_compartment.state_args["left"] = compartment.state_args["val"]
            next_compartment.state_args["right"] = compartment.state_args["val"] + 1
            self.__transition(next_compartment)
            return
        elif __e._message == "Prev":
            next_compartment = None
            next_compartment = StateParamsCompartment('__stateparams_state_Merge', next_compartment)
            next_compartment.state_args["left"] = compartment.state_args["val"] + 1
            next_compartment.state_args["right"] = compartment.state_args["val"]
            self.__transition(next_compartment)
            return
        elif __e._message == "Log":
            self.got_param_do("val",(compartment.state_args["val"]))
            return
    
    # ----------------------------------------
    # $Merge
    
    def __stateparams_state_Merge(self, __e, compartment):
        if __e._message == "Next":
            next_compartment = None
            next_compartment = StateParamsCompartment('__stateparams_state_Split', next_compartment)
            next_compartment.state_args["val"] = compartment.state_args["left"] + compartment.state_args["right"]
            self.__transition(next_compartment)
            return
        elif __e._message == "Prev":
            next_compartment = None
            next_compartment = StateParamsCompartment('__stateparams_state_Split', next_compartment)
            next_compartment.state_args["val"] = compartment.state_args["left"] * compartment.state_args["right"]
            self.__transition(next_compartment)
            return
        elif __e._message == "Log":
            self.got_param_do("left",(compartment.state_args["left"]))
            self.got_param_do("right",(compartment.state_args["right"]))
            return
    
    # ===================== Actions Block =================== #
    
    def got_param_do(self,name: str,val: int):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__stateparams_state_Init':
            self.__stateparams_state_Init(__e, self.__compartment)
        elif self.__compartment.state == '__stateparams_state_Split':
            self.__stateparams_state_Split(__e, self.__compartment)
        elif self.__compartment.state == '__stateparams_state_Merge':
            self.__stateparams_state_Merge(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class StateParamsCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    